### DISCUSSION TITLE
Accessing `.env` variables inside yaml configuration file

### DISCUSSION THREAD

**Discussion by arsenipachkovski** [deployphp/deployer#2683] [Help needed]
Is there a way to access `.env` properties inside `deploy.yaml` configuration file?

As I will commit `deploy.yaml` to my repository, I don't want any private information inside it.

I know that it's possible with `deploy.php` by using dotenv package (e.g. `Dotenv::createImmutable(__DIR__)->load()`), but is there a way with yaml file ?

Here is an example of what I'm trying to achieve with cloudflare's recipe:
```
config:
  ...
  cloudflare:
    - domain: ${CLOUDFLARE_DOMAIN}
    - zone_id: ${CLOUDFLARE_ZONE_ID}
    - email: ${CLOUDFLARE_API_EMAIL}
    - api_key: ${CLOUDFLARE_API_KEY}
```

**Comment by antonmedv** [comment]
Currently no. We can think of it as some feature.

You can create separate task in PHP which will load all needed stuff into config.

**Reply by arsenipachkovski** [reply]
That's what I ended up doing.

Yeah, a feature like that would be really good (and keep yaml file clean).

Thanks for your time.

**Reply by simonbuehler** [reply]
@arsenipachkovski mind sharing your solution?

**Reply by arsenipachkovski** [reply]
What I did is commenting out the variables inside my `deploy.yaml` file (so I know these variables should be defined in my custom recipe + `.env`):
```
config:
  ...
  cloudflare:
    # domain: Defined inside custom recipe
    # zone_id: Defined inside custom recipe
    # email: Defined inside custom recipe
    # api_key: Defined inside custom recipe
```

Then I created a custom recipe file in `deploy/custom.php` and imported it inside my `deploy.yaml`:
```
import:
  - recipe/common.php
  - contrib/cloudflare.php
  - deploy/custom.php <---- HERE
  - ...
```

Inside my `custom.php` recipe, I created a `deploy:env` task which uses `Dotenv` [package](https://github.com/vlucas/phpdotenv) to load environment variables & `add` helper to set environment variables inside `cloudflare` namespace.
```

use Dotenv\Dotenv;

Dotenv::createImmutable(__DIR__ . '/../')->load();

desc('Inject all necessary .env variables inside deployer config');
task('deploy:env', function () {
    add('cloudflare', [
        'zone_id' => env('CLOUDFLARE_ZONE_ID'),
        'email'   => env('CLOUDFLARE_API_EMAIL'),
        'api_key' => env('CLOUDFLARE_API_KEY'),
    ]);
	...
});
```

Finally, in the `tasks` section of my `deploy.yaml` file, I used my custom task to load all my environment variables & make them available to other tasks:
```
tasks:
  deploy:
    - deploy:env
    - ...
```

Hope that helps!

**Reply by simonbuehler** [reply]
great thanks!

**Reply by simonbuehler** [reply]
i just misunderstood that i could set a plain php variable like 
```php
use Dotenv\Dotenv;
$dotenv = Dotenv::createImmutable(dirname(__DIR__, 1), '.env');
$dotenv->load();
$dotenv->required(['DEPLOYER_HOSTNAME', 'DEPLOYER_REMOTE_USER', 'DEPLOYER_PORT', 'DEPLOYER_DEPLOY_PATH', 'DEPLOYER_IDENTITY_FILE'])->notEmpty();

$dep_host = env('DEPLOYER_HOSTNAME');
$dep_port = env('DEPLOYER_PORT'); // a int
```
and then in the yaml assign it like
```yaml
hosts:
  stage:
    hostname:  '{{dep_host}}'
    port:  '{{dep_port}}'
```
but that down't work like that and i couldn't find docs about variable handling

**Reply by antonmedv** [reply]
Read config section of docs

**Reply by simonbuehler** [reply]
You mean docs/config.md ?

**Reply by antonmedv** [reply]
Yes

**Reply by simonbuehler** [reply]
its a empty file https://github.com/deployphp/deployer/blob/master/docs/config.md
